libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,370 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused-parameters function.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template<typename T>
+vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
+{
+	yulAssert(_vec.size() == _mask.size(), "");
+
+	vector<T> ret;
+
+	for (size_t i = 0; i < _mask.size(); ++i)
+		if (_mask[i])
+			ret.push_back(_vec[i]);
+
+	return ret;
+}
+
+/**
+ * Step 1A of UnusedFunctionParameterPruner: Find functions whose return parameters are not used in
+ * the code, i.e.,
+ *
+ * Look at all VariableDeclaration `let x_1, ..., x_i, ... x_n := f(y_1, ..., y_m)` and check if
+ * `x_i` is unused. If all function calls to `f` has its i-th return parameter unused, we will mark
+ * its `i`th index so that it can be removed in later steps.
+ */
+class FindFunctionsWithUnusedReturns: public ASTWalker
+{
+public:
+	explicit FindFunctionsWithUnusedReturns(
+		Dialect const& _dialect,
+		Block const& _block,
+		map<YulString, size_t> const& _references
+	) :
+		m_dialect(_dialect),
+		m_references(_references)
+
+	{
+		(*this)(_block);
+	}
+	using ASTWalker::operator();
+	void operator()(VariableDeclaration const& _v) override;
+
+	map<YulString, vector<bool>> m_returns;
+ private:
+	Dialect const& m_dialect;
+	map<YulString, size_t> const& m_references;
+};
+
+// Find functions whose return parameters are unused. Assuming that the code is in SSA form and that
```

Comment:
Ah, so this is more like removing return parameters that are ignored by the caller?

Do you also remove return parameters that are never assigned in the function?

Reply:
> Do you also remove return parameters that are never assigned in the function?

This is done in Step 1B.

An important case that we are missing is `pop(f())`. A problem with this case is that ExpressionSplitter converts it into

```
let x := f()
pop(x)
```
So we can't infer from `ReferencesCounter` if return value of `f` used. That `pop(x)` is a vacuous statement.

Do you think we should change ExpressionSplitter so that it can `pop(f())` is converted into just `let x := f()`?

Reply:
The `UnusedPruner` should remove `pop(x)`. In general, UnusedPruner should be run before (or at least close to) this step, shouldn't it?

Reply:
This step is getting rather complicated - I didn't review the full thing yet, but is it beneficial to do both things in the same step, or should we rather have one step that only modifies the function definitions (and splits them into two) and another step that modifies the call sites?

Reply:
They can be split up. They both share the same final step, where the functions get renamed, so doing it together makes sense. If we split it up, the `FullInliner` will have to inline two functions, instead of just one.

Reply:
> The UnusedPruner should remove pop(x)

`UnusedPruner` seems to be doing the opposite: converting `let x := f(1)` to `pop(f(1))`. So running `UnusedPruner` means such functions cannot be rewritten.